docs/format_signatures.rst

```diff
@@ -0,0 +1,122 @@
+Formatting signatures
+=====================
+
+This theme provides two alternative ways to automatically format/indent API
+documentation signatures.
+
+CSS-based function parameter wrapping
+-------------------------------------
+
+There is a CSS-based formatting rule that can be enabled for long function
+signatures that displays each function parameter on a separate line.  This is
+enabled by default, and works fairly well for Python signatures.
+
+.. confval:: html_wrap_signatures_with_css
+
+   Specifies for which `Sphinx domains
+   <https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html>`__
+   this CSS-based formatting is enabled.
+
+   Supported types are:
+
+   :py:obj:`bool`
+       If `True`, enable CSS-based formatting for all domains.  If `False`,
+       disable CSS-based formatting for all domains.
+
+   :py:obj:`list[str]`
+       List of domains for which to enable CSS-based formatting.
+
+   The default value is :python:`True`
+
+.. confval:: html_wrap_signatures_with_css_column_limit
+
+   Maximum signature length before function parameters are displayed on separate
+   lines.
+
+   Only applies to domains for which :confval:`html_wrap_signatures_with_css` is
+   enabled.
+
+   The default value is :python:`70`.
+
+.. rst-example::
+
+   .. py:function:: long_function_signature_example(\
+                      name: int, \
+                      other_param: list[str], \
+                      yet_another: bool = False, \
+                      finally_another: str = "some long string goes here"\
+                    ) -> Tuple[A, Very, Long, Return, Type]
+
+clang-format-based function parameter wrapping
+-----------------------------------------------
+
+There is a more powerful alternative formatting mechanism based on `clang-format
+<https://clang.llvm.org/docs/ClangFormat.html>`__.  This supports C, C++, Java,
+JavaScript, Objective-C, and C#.
+
+This functionality is available as a separate extension included with this
+theme.  To use it, you must include it in your conf.py file and you must also
+set the :confval:`clang_format_signatures_domain_styles` configuration option.
+
+.. code-block:: python
+
+    extensions = [
+        # other extensions...
+        "sphinx_immaterial.format_signatures",
+    ]
+
+    clang_format_signatures_domain_styles = {
+        "cpp": """{
+        BasedOnStyle: LLVM,
+        ColumnLimit: 70,
+        }""",
```

Comment:
Are the curly brackets necessary here?

I understand this is a documented example, but my question is more specific to the extension.

Reply:
This is just exactly following the syntax required by the clang-format `-style` command-line option.

Allowed values are:

- `file`
- `LLVM`
- `GNU`
- `Google`
- `Chromium`
- `Microsoft`
- `Mozilla`
- `WebKit`
- `{key:value, ...}`

Requiring the user to include the curly braces in the style means they could theoretically use one of the other plain-string values, and makes it consistent with the clang-format command-line syntax.

However, it is true that given the recommendation to use a column limit of 68 (previously 70), none of these options except the last one is likely to be used, so we could just add the curly braces.

Reply:
Technically speaking the curly brackets are an implementation detail that I don't think should be burdened on the user. It might be better to encourage `LLVM`, `GNU`, `Google`, `Chromium`, `Microsoft`, `Mozilla`, `WebKit`, or `file` (or `file:path/relative/to/docs/src/.clang-format` for v14+). That way a complex style definition is properly abstracted from being in a dict's key's value. If users are going to define their own style definition, then it will not likely be as simple as a few lines (based on my experience and recommendations from others).

Coincidentally, I also coded this `subprocess.run()` call identically [in a github action I re-wrote in python](https://github.com/shenxianpeng/cpp-linter-action/blob/9f2caa8fa2e3e7a5bf39103e634a7d754c36c056/cpp_linter/run.py#L455). However, that action explicitly supports the predefined options (which can be clang version dependent) & implicitly supports the curly bracket option if the user is clever enough (though its not at all encouraged).

Reply:
I'm not sure I understand what you are proposing --- if we always surround the user's style with curly braces, then the user can't specify one of those other options, like LLVM, file, etc.

Reply:
I'm ok with leaving it as is because adding curly brackets would be only apply if a predefined style isn't used. So if you wanted to change it, then there may need to be a 
```py
if style not in (
    "LLVM", "GNU", "Google", "Chromium", "Microsoft", "Mozilla", "WebKit"
) or not style.startswith("file"):
    style = "{" + style + "}"
```
But, that would less sustainable because clang-format may choose to add or remove a predefined style at any time in the future.

By encouraging the use of predefined styles (or `file`), then we might be able to avoid testing for predefined styles and still implicitly support inline curly bracket styles.

Reply:
The thing is a predefined style will not have the right column limit --- and similarly the project's own .clang-format also won't have the right column limit.

Reply:
Actually, some styles do define a `ColumnLimit`. It sounds like the preferred method is to use custom styles. If curly brackets are not going to be automatic, then a note about that might help (since the YAML syntax doesn't really use curly brackets the same as JSON).

Reply:
Yeah I think they all specify a column limit, but they don't use a column limit of 68, which is what works best for this theme.  Probably they mostly use 80.

I added a note about curly braces being required.

Note that the style syntax is not yaml, but rather a custom syntax used by clang-format.

Reply:
From clang-format docs:
> The `.clang-format` file uses YAML format

But maybe there's a difference in .clang-format file vs CLI value.

Reply:
Oh, I think you are right -- sorry.